great_expectations/datasource/fluent/sql_datasource.py

```diff
@@ -712,6 +723,71 @@ def build_batch_request(
             partitioner=partitioner,
         )
 
+    @override
+    def add_batch_definition(
+        self,
+        name: str,
+        partitioner: Optional[ColumnPartitioner] = None,
+    ) -> BatchDefinition[ColumnPartitioner]:
+        if partitioner:
+            self.validate_batch_definition(partitioner)
+        return super().add_batch_definition(name, partitioner)
+
+    @public_api
+    def validate_batch_definition(self, partitioner: ColumnPartitioner) -> None:
+        """Validates that the batch definition column is of a permissible type
+
+        This isn't meant to be called directly. This is called internally when a batch definition
+         is added. Data asset implementers can override this for their specific data asset.
+
+        Raises:
+            SqlAddBatchDefinitionError: The specified column to partition on is not of
+                a permissible type for batching (ie date or datetime) or no data is
+                present in this column.
+        """
+        # We only support certain partitioners for using as batch definitions.
+        assert isinstance(
+            partitioner,
+            (
+                ColumnPartitionerYearly,
+                ColumnPartitionerMonthly,
+                ColumnPartitionerDaily,
+            ),
+        )
+        # A _SQLAsset must have a SQLDatasource
```

Comment:
I'm not sur how to enforce since with how we currently have typing setup in this class.

Reply:
Yeah, I think in general we could make better use of generics so this wasn't necessary, but agreed that it's needed in the current state.